File: src/render.js
import { c, truncate, fmtDuration, fmtNum, pad } from "./util.js";
import { toolOneLiner, stats, filesTouched, userTurns } from "./model.js";

const DEFAULTS = { thinking: false, full: false, sidechains: false, maxLines: 12, maxChars: 4000, width: 100 };

function indent(text, prefix) {
  return String(text)
    .split("\n")
    .map((l) => prefix + l)
    .join("\n");
}

function clip(text, o) {
  if (o.full) return text;
  let lines = String(text ?? "").split("\n");
  let cut = false;
  if (lines.length > o.maxLines) {
    lines = lines.slice(0, o.maxLines);
    cut = true;
  }
  let out = lines.join("\n");
  if (out.length > o.maxChars) {
    out = out.slice(0, o.maxChars);
    cut = true;
  }
  return cut ? out + `\n${c.dim}… (truncated; use --full)${c.reset}` : out;
}

function clock(ev, start) {
  const t = Date.parse(ev.ts) - start;
  return Number.isNaN(t) ? "        " : pad("+" + fmtDuration(t), 8);
}

/** Render one event to a terminal string (may be multi-line). Returns null when hidden. */
export function formatEvent(ev, opts = {}, ctx = {}) {
  const o = { ...DEFAULTS, ...opts };
  if (ev.sidechain && !o.sidechains) return null;
  const start = ctx.start ?? Date.parse(ev.ts);
  const tag = `${c.gray}${clock(ev, start)}${c.reset} `;
  const side = ev.sidechain ? `${c.magenta}[subagent] ${c.reset}` : "";
  switch (ev.kind) {
    case "user":
      return `\n${tag}${c.bold}${c.cyan}▶ user (turn ${ev.turn})${c.reset}\n${indent(ev.text, "  ")}\n`;
    case "assistant":
      return `${tag}${side}${c.green}●${c.reset}${ev.model ? ` ${c.dim}${ev.model}${c.reset}` : ""}\n${indent(clip(ev.text, o), "  ")}`;
    case "thinking":
      if (!o.thinking) return null;
      if (!ev.text?.trim()) return `${tag}${side}${c.gray}∴ thinking (hidden by provider)${c.reset}`;
      return `${tag}${side}${c.gray}∴ thinking${c.reset}\n${c.gray}${indent(clip(ev.text, o), "  ")}${c.reset}`;
    case "tool_call":
      return `${tag}${side}${c.yellow}⚙ ${toolOneLiner(ev, o.full ? 100000 : 160)}${c.reset}`;
    case "tool_result": {
      const err = ev.result?.isError;
      const body = clip(ev.result?.output ?? "", { ...o, maxLines: o.full ? 1e9 : Math.min(o.maxLines, 6) });
      if (!body.trim()) return `${tag}${side}${c.dim}  ↳ (empty result)${c.reset}`;
      return `${(err ? c.red : c.dim)}${indent(body, "    │ ")}${c.reset}`;
    }
    case "system":
      return `${tag}${c.blue}◇ ${ev.subtype || "system"}${c.reset} ${c.dim}${truncate(String(ev.text ?? "").replace(/\s+/g, " "), 160)}${c.reset}`;
    case "error":
      return `${tag}${c.red}✖ ${truncate(ev.text, 500)}${c.reset}`;
    default:
      return null;
  }
}

export function renderHeader(session) {
  const s = stats(session);
  const lines = [];
  lines.push(`${c.bold}${session.harness}${c.reset} session ${c.dim}${session.id}${c.reset}`);
  if (session.title) lines.push(`  title:    ${session.title}`);
  if (session.model) lines.push(`  model:    ${session.model}`);
  if (session.cwd) lines.push(`  cwd:      ${session.cwd}${session.gitBranch ? ` (${session.gitBranch}${session.gitCommit ? "@" + session.gitCommit.slice(0, 8) : ""})` : ""}`);
  if (session.startedAt) lines.push(`  started:  ${session.startedAt}  duration: ${fmtDuration(s.durationMs)}`);
  lines.push(`  turns: ${s.turns}  assistant msgs: ${s.assistantMessages}  tool calls: ${s.toolCalls} (${s.toolErrors} errors)  files touched: ${s.filesTouched}`);
  lines.push(`  tokens: in ${fmtNum(s.usage.input)}  out ${fmtNum(s.usage.output)}  cache read ${fmtNum(s.usage.cacheRead)}${s.costUsd != null ? `  cost $${s.costUsd.toFixed(4)}` : ""}`);
  if (session.path) lines.push(`  ${c.dim}${session.path}${c.reset}`);
  return lines.join("\n");
}

export function renderTranscript(session, opts = {}) {
  const start = Date.parse(session.startedAt || session.events[0]?.ts);
  const out = [renderHeader(session), ""];
  for (const ev of session.events) {
    if (opts.turn && ev.turn !== opts.turn) continue;
    const s = formatEvent(ev, opts, { start });
    if (s != null) out.push(s);
  }
  return out.join("\n");
}

export function renderStats(session) {
  const s = stats(session);
  const rows = [
    ["harness", s.harness],
    ["model", s.model || "-"],
    ["turns", s.turns],
    ["assistant messages", s.assistantMessages],
    ["thinking blocks", s.thinkingBlocks],
    ["tool calls", s.toolCalls],
    ["tool errors", s.toolErrors],
    ["files touched", s.filesTouched],
    ["duration", fmtDuration(s.durationMs)],
    ["input tokens", fmtNum(s.usage.input)],
    ["output tokens", fmtNum(s.usage.output)],
    ["cache read tokens", fmtNum(s.usage.cacheRead)],
    ["cache write tokens", fmtNum(s.usage.cacheWrite)],
    ["cost (USD)", s.costUsd != null ? s.costUsd.toFixed(4) : "-"],
  ];
  const w = Math.max(...rows.map((r) => r[0].length));
  const lines = rows.map(([k, v]) => `${pad(k, w)}  ${v}`);
  lines.push("", "tools by name:");
  for (const [n, k] of Object.entries(s.toolsByName).sort((a, b) => b[1] - a[1])) lines.push(`  ${pad(n, 24)} ${k}`);
  const files = filesTouched(session);
  if (files.length) {
    lines.push("", "files touched:");
    for (const f of files) lines.push(`  ${f.path}  (${f.ops.join(",")})`);
  }
  return lines.join("\n");
}

/** Markdown export of a session. */
export function renderMarkdown(session, opts = {}) {
  const o = { ...DEFAULTS, maxLines: 40, ...opts };
  const s = stats(session);
  const md = [];
  md.push(`# ${session.title || session.id}`, "");
  md.push(`- harness: ${session.harness}`);
  md.push(`- session: ${session.id}`);
  if (session.model) md.push(`- model: ${session.model}`);
  if (session.cwd) md.push(`- cwd: ${session.cwd}${session.gitBranch ? ` (${session.gitBranch})` : ""}`);
  if (session.startedAt) md.push(`- started: ${session.startedAt}, duration ${fmtDuration(s.durationMs)}`);
  md.push(`- turns: ${s.turns}, tool calls: ${s.toolCalls}, tokens in/out: ${fmtNum(s.usage.input)}/${fmtNum(s.usage.output)}`);
  md.push("");
  for (const ev of session.events) {
    if (ev.sidechain && !o.sidechains) continue;
    switch (ev.kind) {
      case "user":
        md.push(`## Turn ${ev.turn} — user`, "", quote(ev.text), "");
        break;
      case "assistant":
        md.push(`**assistant**${ev.model ? ` (${ev.model})` : ""}:`, "", ev.text, "");
        break;
      case "thinking":
        if (o.thinking && ev.text?.trim()) md.push(`<details><summary>thinking</summary>`, "", ev.text, "", `</details>`, "");
        break;
      case "tool_call":
        md.push(`- 🔧 \`${toolOneLiner(ev, 200).replace(/`/g, "'")}\``);
        break;
      case "tool_result": {
        const body = o.full ? ev.result.output : clip(ev.result.output, { ...o, maxLines: 20 }).replace(/\x1b\[[0-9;]*m/g, "");
        if (body.trim()) md.push("", "  ```", indent(body, "  "), "  ```", "");
        break;
      }
      case "system":
        md.push(`> _${ev.subtype}_: ${truncate(String(ev.text ?? "").replace(/\s+/g, " "), 200)}`, "");
        break;
      case "error":
        md.push(`> ❌ ${ev.text}`, "");
        break;
    }
  }
  return md.join("\n");
}

function quote(text) {
  return String(text)
    .split("\n")
    .map((l) => "> " + l)
    .join("\n");
}

export function renderSessionList(items) {
  if (!items.length) return "(no sessions found)";
  const lines = [];
  lines.push(`${c.bold}${pad("harness", 12)}${pad("id", 38)}${pad("updated", 21)}${pad("cwd", 34)}title${c.reset}`);
  for (const it of items) {
    const cwd = it.cwd ? truncate(it.cwd.replace(process.env.HOME || "", "~"), 32) : "";
    lines.push(`${pad(it.harness, 12)}${pad(it.id, 38)}${pad((it.updatedAt || "").slice(0, 19).replace("T", " "), 21)}${pad(cwd, 34)}${truncate(it.title || "", 60)}`);
  }
  return lines.join("\n");
}

export { userTurns };

File: src/render.rs
//! Terminal, markdown, and list renderers.
use crate::adapters::SessionSummary;
use crate::model::{files_touched, sorted_counts, stats, tool_one_liner, Event, EventKind, Session};
use crate::util::{colors, fmt_duration, fmt_num, home_dir, indent, one_line, pad, truncate, ts_ms};

#[derive(Clone, Debug)]
pub struct RenderOpts {
    pub thinking: bool,
    pub full: bool,
    pub sidechains: bool,
    pub max_lines: usize,
    pub max_chars: usize,
    pub turn: Option<u32>,
}

impl Default for RenderOpts {
    fn default() -> Self {
        RenderOpts { thinking: false, full: false, sidechains: false, max_lines: 12, max_chars: 4000, turn: None }
    }
}

fn clip(text: &str, o: &RenderOpts, max_lines: usize) -> String {
    if o.full {
        return text.to_string();
    }
    let c = colors();
    let mut lines: Vec<&str> = text.lines().collect();
    let mut cut = false;
    if lines.len() > max_lines {
        lines.truncate(max_lines);
        cut = true;
    }
    let mut out = lines.join("\n");
    if out.chars().count() > o.max_chars {
        out = out.chars().take(o.max_chars).collect();
        cut = true;
    }
    if cut {
        out.push_str(&format!("\n{}… (truncated; use --full){}", c.dim, c.reset));
    }
    out
}

fn clock(ev: &Event, start: Option<i64>) -> String {
    match (ts_ms(&ev.ts), start) {
        (Some(t), Some(s)) => pad(&format!("+{}", fmt_duration(t - s)), 8),
        _ => "        ".into(),
    }
}

/// Render one event (may be multi-line). None when hidden by options.
pub fn format_event(ev: &Event, o: &RenderOpts, start: Option<i64>) -> Option<String> {
    if ev.sidechain && !o.sidechains {
        return None;
    }
    let c = colors();
    let tag = format!("{}{}{} ", c.gray, clock(ev, start), c.reset);
    let side = if ev.sidechain { format!("{}[subagent] {}", c.magenta, c.reset) } else { String::new() };
    let text = ev.text_str();
    Some(match ev.kind {
        EventKind::User => format!("\n{tag}{}{}▶ user (turn {}){}\n{}\n", c.bold, c.cyan, ev.turn, c.reset, indent(text, "  ")),
        EventKind::Assistant => {
            let model = ev.model.as_ref().map(|m| format!(" {}{}{}", c.dim, m, c.reset)).unwrap_or_default();
            format!("{tag}{side}{}●{}{model}\n{}", c.green, c.reset, indent(&clip(text, o, o.max_lines), "  "))
        }
        EventKind::Thinking => {
            if !o.thinking {
                return None;
            }
            if text.trim().is_empty() {
                format!("{tag}{side}{}∴ thinking (hidden by provider){}", c.gray, c.reset)
            } else {
                format!("{tag}{side}{}∴ thinking{}\n{}{}{}", c.gray, c.reset, c.gray, indent(&clip(text, o, o.max_lines), "  "), c.reset)
            }
        }
        EventKind::ToolCall => format!("{tag}{side}{}⚙ {}{}", c.yellow, tool_one_liner(ev, if o.full { 100_000 } else { 160 }), c.reset),
        EventKind::ToolResult => {
            let r = ev.result.as_ref()?;
            let body = clip(&r.output, o, if o.full { usize::MAX } else { o.max_lines.min(6) });
            if body.trim().is_empty() {
                format!("{tag}{side}{}  ↳ (empty result){}", c.dim, c.reset)
            } else {
                format!("{}{}{}", if r.is_error { c.red } else { c.dim }, indent(&body, "    │ "), c.reset)
            }
        }
        EventKind::System => format!("{tag}{}◇ {}{} {}{}{}", c.blue, ev.subtype.as_deref().unwrap_or("system"), c.reset, c.dim, truncate(&one_line(text), 160), c.reset),
        EventKind::Error => format!("{tag}{}✖ {}{}", c.red, truncate(text, 500), c.reset),
    })
}

pub fn session_start_ms(session: &Session) -> Option<i64> {
    session.started_at.as_deref().and_then(ts_ms).or_else(|| session.events.first().and_then(|e| ts_ms(&e.ts)))
}

pub fn render_header(session: &Session) -> String {
    let c = colors();
    let s = stats(session);
    let mut lines = vec![format!("{}{}{} session {}{}{}", c.bold, session.harness(), c.reset, c.dim, session.id, c.reset)];
    if let Some(t) = &session.title {
        lines.push(format!("  title:    {t}"));
    }
    if let Some(m) = &session.model {
        lines.push(format!("  model:    {m}"));
    }
    if let Some(cwd) = &session.cwd {
        let git = match (&session.git_branch, &session.git_commit) {
            (Some(b), Some(cm)) => format!(" ({b}@{})", &cm[..cm.len().min(8)]),
            (Some(b), None) => format!(" ({b})"),
            _ => String::new(),
        };
        lines.push(format!("  cwd:      {cwd}{git}"));
    }
    if let Some(st) = &session.started_at {
        lines.push(format!("  started:  {st}  duration: {}", fmt_duration(s.duration_ms)));
    }
    lines.push(format!(
        "  turns: {}  assistant msgs: {}  tool calls: {} ({} errors)  files touched: {}",
        s.turns, s.assistant_messages, s.tool_calls, s.tool_errors, s.files_touched
    ));
    let cost = s.cost_usd.map(|c| format!("  cost ${c:.4}")).unwrap_or_default();
    lines.push(format!("  tokens: in {}  out {}  cache read {}{}", fmt_num(s.usage.input), fmt_num(s.usage.output), fmt_num(s.usage.cache_read), cost));
    if let Some(p) = &session.path {
        lines.push(format!("  {}{}{}", c.dim, p, c.reset));
    }
    lines.join("\n")
}

pub fn render_transcript(session: &Session, o: &RenderOpts) -> String {
    let start = session_start_ms(session);
    let mut out = vec![render_header(session), String::new()];
    for ev in &session.events {
        if o.turn.is_some_and(|t| ev.turn != t) {
            continue;
        }
        if let Some(s) = format_event(ev, o, start) {
            out.push(s);
        }
    }
    out.join("\n")
}

pub fn render_stats(session: &Session) -> String {
    let s = stats(session);
    let rows: Vec<(&str, String)> = vec![
        ("harness", session.harness().to_string()),
        ("model", s.model.clone().unwrap_or_else(|| "-".into())),
        ("turns", s.turns.to_string()),
        ("assistant messages", s.assistant_messages.to_string()),
        ("thinking blocks", s.thinking_blocks.to_string()),
        ("tool calls", s.tool_calls.to_string()),
        ("tool errors", s.tool_errors.to_string()),
        ("files touched", s.files_touched.to_string()),
        ("duration", fmt_duration(s.duration_ms)),
        ("input tokens", fmt_num(s.usage.input)),
        ("output tokens", fmt_num(s.usage.output)),
        ("cache read tokens", fmt_num(s.usage.cache_read)),
        ("cache write tokens", fmt_num(s.usage.cache_write)),
        ("cost (USD)", s.cost_usd.map(|c| format!("{c:.4}")).unwrap_or_else(|| "-".into())),
    ];
    let w = rows.iter().map(|r| r.0.len()).max().unwrap_or(0);
    let mut lines: Vec<String> = rows.iter().map(|(k, v)| format!("{}  {v}", pad(k, w))).collect();
    lines.push(String::new());
    lines.push("tools by name:".into());
    for (n, k) in sorted_counts(&s.tools_by_name) {
        lines.push(format!("  {} {k}", pad(&n, 24)));
    }
    let files = files_touched(session);
    if !files.is_empty() {
        lines.push(String::new());
        lines.push("files touched:".into());
        for f in files {
            lines.push(format!("  {}  ({})", f.path, f.ops.join(",")));
        }
    }
    lines.join("\n")
}

fn quote(text: &str) -> String {
    text.lines().map(|l| format!("> {l}")).collect::<Vec<_>>().join("\n")
}

/// Markdown export of a session.
pub fn render_markdown(session: &Session, o: &RenderOpts) -> String {
    let o = RenderOpts { max_lines: 40, ..o.clone() };
    let s = stats(session);
    let mut md: Vec<String> = Vec::new();
    md.push(format!("# {}", session.title.clone().unwrap_or_else(|| session.id.clone())));
    md.push(String::new());
    md.push(format!("- harness: {}", session.harness()));
    md.push(format!("- session: {}", session.id));
    if let Some(m) = &session.model {
        md.push(format!("- model: {m}"));
    }
    if let Some(cwd) = &session.cwd {
        md.push(format!("- cwd: {cwd}{}", session.git_branch.as_ref().map(|b| format!(" ({b})")).unwrap_or_default()));
    }
    if let Some(st) = &session.started_at {
        md.push(format!("- started: {st}, duration {}", fmt_duration(s.duration_ms)));
    }
    md.push(format!("- turns: {}, tool calls: {}, tokens in/out: {}/{}", s.turns, s.tool_calls, fmt_num(s.usage.input), fmt_num(s.usage.output)));
    md.push(String::new());
    for ev in &session.events {
        if ev.sidechain && !o.sidechains {
            continue;
        }
        let text = ev.text_str();
        match ev.kind {
            EventKind::User => {
                md.push(format!("## Turn {} — user", ev.turn));
                md.push(String::new());
                md.push(quote(text));
                md.push(String::new());
            }
            EventKind::Assistant => {
                md.push(format!("**assistant**{}:", ev.model.as_ref().map(|m| format!(" ({m})")).unwrap_or_default()));
                md.push(String::new());
                md.push(text.to_string());
                md.push(String::new());
            }
            EventKind::Thinking => {
                if o.thinking && !text.trim().is_empty() {
                    md.extend(["<details><summary>thinking</summary>".into(), String::new(), text.to_string(), String::new(), "</details>".into(), String::new()]);
                }
            }
            EventKind::ToolCall => md.push(format!("- 🔧 `{}`", tool_one_liner(ev, 200).replace('`', "'"))),
            EventKind::ToolResult => {
                if let Some(r) = &ev.result {
                    let body = if o.full { r.output.clone() } else { clip(&r.output, &o, 20) };
                    let body = strip_ansi(&body);
                    if !body.trim().is_empty() {
                        md.extend([String::new(), "  ```".into(), indent(&body, "  "), "  ```".into(), String::new()]);
                    }
                }
            }
            EventKind::System => {
                md.push(format!("> _{}_: {}", ev.subtype.as_deref().unwrap_or("system"), truncate(&one_line(text), 200)));
                md.push(String::new());
            }
            EventKind::Error => {
                md.push(format!("> ❌ {text}"));
                md.push(String::new());
            }
        }
    }
    md.join("\n")
}

fn strip_ansi(s: &str) -> String {
    let mut out = String::new();
    let mut chars = s.chars().peekable();
    while let Some(ch) = chars.next() {
        if ch == '\x1b' && chars.peek() == Some(&'[') {
            chars.next();
            for c in chars.by_ref() {
                if c.is_ascii_alphabetic() {
                    break;
                }
            }
        } else {
            out.push(ch);
        }
    }
    out
}

pub fn render_session_list(items: &[SessionSummary]) -> String {
    if items.is_empty() {
        return "(no sessions found)".into();
    }
    let c = colors();
    let home = home_dir().display().to_string();
    let mut lines = vec![format!("{}{}{}{}{}title{}", c.bold, pad("harness", 12), pad("id", 38), pad("updated", 21), pad("cwd", 34), c.reset)];
    for it in items {
        let cwd = it.cwd.as_deref().map(|d| truncate(&d.replacen(&home, "~", 1), 32)).unwrap_or_default();
        let updated = it.updated_at.chars().take(19).collect::<String>().replace('T', " ");
        lines.push(format!("{}{}{}{}{}", pad(it.harness.as_str(), 12), pad(&it.id, 38), pad(&updated, 21), pad(&cwd, 34), truncate(&it.title, 60)));
    }
    lines.join("\n")
}
